File: components/utils/rbac.ts
//components/utils/rbac.ts
import { Role, User, Plant } from '../../types';

const overlap = (a?: string[], b?: string[]) =>
  !!(a && b && a.some(x => b.includes(x)));

export const canViewUser = (actor: User, target: User) => {
  const ar = actor.role, tr = target.role;
  if (ar === Role.ADMIN) return true;
  if (ar === Role.OPERATOR) return true;
  if (ar === Role.COORDINATOR) {
    if (tr === Role.ADMIN) return false;
    if ([Role.SUPERVISOR,Role.TECHNICIAN,Role.ASSISTANT].includes(tr))
      return overlap(actor.plantIds, target.plantIds);
    return true;
  }
  if (ar === Role.SUPERVISOR) {
    if ([Role.ADMIN,Role.COORDINATOR].includes(tr)) return false;
    if ([Role.TECHNICIAN,Role.ASSISTANT].includes(tr))
      return overlap(actor.plantIds, target.plantIds);
    return (tr === Role.SUPERVISOR && actor.id === target.id);
  }
  if (ar === Role.TECHNICIAN) {
    if (tr === Role.ASSISTANT) return overlap(actor.plantIds, target.plantIds);
    return (tr === Role.TECHNICIAN && actor.id === target.id);
  }
  if (ar === Role.ASSISTANT) {
    return tr === Role.TECHNICIAN && overlap(actor.plantIds, target.plantIds);
  }
  return false;
};

export const canEditUser = (actor: User, target: User) => {
  const ar = actor.role, tr = target.role;
  if (ar === Role.ADMIN) return true;
  if (ar === Role.OPERATOR) return [Role.OPERATOR,Role.TECHNICIAN,Role.ASSISTANT].includes(tr);
  if (ar === Role.COORDINATOR)
    return tr !== Role.ADMIN && [Role.SUPERVISOR,Role.TECHNICIAN,Role.ASSISTANT].includes(tr)
           && overlap(actor.plantIds, target.plantIds);
  if (ar === Role.SUPERVISOR)
    return ![Role.ADMIN,Role.COORDINATOR].includes(tr)
           && [Role.TECHNICIAN,Role.ASSISTANT].includes(tr)
           && overlap(actor.plantIds, target.plantIds);
  if (ar === Role.TECHNICIAN)
    return tr === Role.ASSISTANT && overlap(actor.plantIds, target.plantIds);
  return false;
};

export const canViewPlant = (actor: User, plant: Plant) =>
  actor.role === Role.ADMIN || actor.role === Role.OPERATOR || (actor.plantIds || []).includes(plant.id);

export const canEditPlant = (actor: User, plant: Plant) =>
  actor.role === Role.ADMIN || actor.role === Role.OPERATOR ||
  ([Role.COORDINATOR,Role.SUPERVISOR].includes(actor.role) && (actor.plantIds || []).includes(plant.id));

File: attachments/app/core/rbac.py
# LOOPOS/attachments/app/core/rbac.py
from typing import List, Dict

#Role = str  # usando seus literais em pt-br

def _overlap(a: List[str], b: List[str]) -> bool:
    return bool(set(a or []) & set(b or []))


def can_view_user(actor: Dict, target: Dict) -> bool:
    # ✅ Qualquer um vê a si mesmo
    if actor["id"] == target["id"]:
        return True
    
    ar, tr = actor["role"], target["role"]


    if ar == "Admin": 
        return True
    if ar == "Operador":
        return tr in {"Operador","Técnico","Auxiliar","Supervisor","Coordenador","Admin"}
    if ar == "Coordenador":
        if tr in {"Admin"}: return False
        if tr in {"Supervisor","Técnico","Auxiliar"}:
            return _overlap(actor.get("plantIds",[]), target.get("plantIds",[]))
        return True
    if ar == "Supervisor":
        if tr in {"Admin","Coordenador"}: return False
        if tr in {"Técnico","Auxiliar"}:
            actor_plants = actor.get("plantIds",[])
            target_plants = target.get("plantIds",[])
            result = _overlap(actor_plants, target_plants)
            

            return result
        if tr == "Supervisor":
            return _overlap(actor.get("plantIds",[]), target.get("plantIds",[]))
        return False
    if ar == "Técnico":
        if tr == "Auxiliar":
            return _overlap(actor.get("plantIds",[]), target.get("plantIds",[]))
        return tr == "Técnico" and actor["id"] == target["id"]
    if ar == "Auxiliar":
        return tr == "Técnico" and _overlap(actor.get("plantIds",[]), target.get("plantIds",[]))
    return False



def can_edit_user(actor: Dict, target: Dict) -> bool:
    ar, tr = actor["role"], target["role"]
    
    # ✅ Qualquer um edita a si mesmo
    if actor["id"] == target["id"]:
        return True
    
    if ar == "Admin":
        return True
    if ar == "Operador":
        return tr in {"Operador","Técnico","Auxiliar"}
    if ar == "Coordenador":
        if tr in {"Admin"}: return False
        if tr in {"Supervisor","Técnico","Auxiliar"}:
            return _overlap(actor.get("plantIds",[]), target.get("plantIds",[]))
        return False
    if ar == "Supervisor":
        if tr in {"Admin","Coordenador"}: return False
        if tr in {"Técnico","Auxiliar"}:
            return _overlap(actor.get("plantIds",[]), target.get("plantIds",[]))
        return False
    if ar == "Técnico":
        return tr == "Auxiliar" and _overlap(actor.get("plantIds",[]), target.get("plantIds",[]))
    return False


def can_view_plant(actor: Dict, plant_id: str) -> bool:
    if actor["role"] in {"Admin","Operador"}: 
        return True
    return plant_id in (actor.get("plantIds") or [])


def can_edit_plant(actor: Dict, plant_id: str) -> bool:
    if actor["role"] in {"Admin","Operador"}: 
        return True
    if actor["role"] in {"Coordenador","Supervisor"}:
        return plant_id in (actor.get("plantIds") or [])
    return False